akka-projection-r2dbc/src/main/scala/akka/projection/r2dbc/internal/R2dbcOffsetStore.scala

```diff
@@ -806,43 +906,54 @@ private[projection] class R2dbcOffsetStore(
   }
 
   def deleteOldTimestampOffsets(): Future[Long] = {
-    if (idle.getAndSet(true)) {
-      // no new offsets stored since previous delete
-      Future.successful(0)
-    } else {
-      val currentState = getState()
-      if (!triggerDeletion.getAndSet(false) && currentState.window.compareTo(settings.timeWindow) < 0) {
-        // it hasn't filled up the window yet
-        Future.successful(0)
-      } else {
-        val until = currentState.latestTimestamp.minus(settings.timeWindow)
-
-        val notInLatestBySlice = currentState.latestBySlice.collect {
-          case record if record.timestamp.isBefore(until) =>
-            // note that deleteOldTimestampOffsetSql already has `AND timestamp_offset < ?`
-            // and that's why timestamp >= until don't have to be included here
-            LatestBySlice(record.slice, record.pid, record.seqNr)
-        }
-        val result = dao.deleteOldTimestampOffset(until, notInLatestBySlice)
-        result.failed.foreach { exc =>
-          idle.set(false) // try again next tick
-          logger.warn(
-            "Failed to delete timestamp offset until [{}] for projection [{}]: {}",
-            until,
-            projectionId.id,
-            exc.toString)
+    // This is running in the background, so fine to progress slowly one slice at a time
+    def loop(slice: Int, count: Long): Future[Long] = {
+      if (slice > maxSlice)
+        Future.successful(count)
+      else
+        deleteOldTimestampOffsets(slice).flatMap { c =>
+          loop(slice + 1, count + c)
         }
-        if (logger.isDebugEnabled)
-          result.foreach { rows =>
-            logger.debug(
-              "Deleted [{}] timestamp offset rows until [{}] for projection [{}].",
-              rows,
-              until,
-              projectionId.id)
-          }
+    }
 
-        result
+    val result = loop(minSlice, 0L)
+
+    if (logger.isDebugEnabled)
+      result.foreach { rows =>
+        logger.debug("{} Deleted [{}] timestamp offset rows", logPrefix, rows)
       }
+
+    result.andThen(_ => scheduleNextDelete())
+  }
+
+  def deleteOldTimestampOffsets(slice: Int): Future[Long] = {
+    val triggerDeletion = triggerDeletionPerSlice.put(slice, FALSE)
+    val currentState = getState()
+    if ((triggerDeletion == null || triggerDeletion == TRUE) && currentState.bySliceSorted.contains(slice)) {
```

Comment:
With booleans in the map, `triggerDeletion` should default to `false` instead of possibly `null`? So can be simplified?

```suggestion
    if (triggerDeletion && currentState.bySliceSorted.contains(slice)) {
```

Reply:
Then it's not the same thing for when the slice is not in the map. The intention is to trigger a deletion for a slice also when it has not been added to the map. It's added to the map when saving offset, but there could still be things to delete in the db even though no new offsets are saved for the slice.

Reply:
Ah, I did misread that. Makes sense if it's also checking presence, and the use of the boxed `java.lang.Boolean`.